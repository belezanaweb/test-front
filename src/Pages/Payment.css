.payment-form {
  background: white;
  border-radius: 3px;
  box-shadow: 0 0 3px rgba(35, 26, 26, 0.4) ;
  padding: 12px;
}
