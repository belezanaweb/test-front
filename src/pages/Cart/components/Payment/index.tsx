import React, { useRef, useCallback, useState, useContext } from 'react';
import * as Yup from 'yup';
import Cards, { Focused } from 'react-credit-cards';
import 'react-credit-cards/es/styles-compiled.css';
import { Form } from '@unform/web';
import { FormHandles } from '@unform/core';
import { useNavigate } from 'react-router';
import { useDispatch } from 'react-redux';

import getValidationError from '../../../../helpers/validations';
import {
  CARD_NUMBER_PLACEHOLDER,
  CVV_PLACEHOLDER,
  DATE_PLACEHOLDER,
  TITULAR_NAME_PLACEHOLDER
} from '../../../../constants/placeholder';

import Button from '../../../../components/Button';
import Input from '../../../../components/Input';
import SumInfo from '../../../../components/SumInfo';

import { Container, FormContent, FormGroup, Content, InputsContent, CartContent } from './styles';

import { setCreditCardInfo } from '../../../../store/modules/cart/actions';
import { setToLocalStorage } from '../../../../helpers/local-storage';
import { BELEZA_NA_WEB_CREDIT_CARD } from '../../../../constants/local-storage';
import { useCartContext } from '../..';

interface PaymentInfo {
  cardNumber: string;
  titularName: string;
  validate: string;
  cvv: string;
  focused: Focused;
}

export default function Payment() {
  const formRef = useRef<FormHandles>(null);
  const { cartItems, creditCardInfo } = useCartContext();
  const [paymentInfo, setPaymentInfo] = useState<PaymentInfo>({} as PaymentInfo);

  const navigate = useNavigate();
  const dispatch = useDispatch();

  const validForm = async () => {
    try {
      const data = formRef?.current?.getData();

      formRef.current?.setErrors({});

      const schema = Yup.object().shape({
        cardNumber: Yup.string().required('Digite o número do cartão'),
        titularName: Yup.string().required('Digite o nome do titular'),
        validate: Yup.string().required('Digite a validade do cartão'),
        cvv: Yup.string().required('Digite o código do cartão')
      });

      await schema.validate(data, { abortEarly: false });

      return true;
    } catch (err) {
      if (err instanceof Yup.ValidationError) {
        const errors = getValidationError(err);
        formRef.current?.setErrors(errors);
        return;
      }

      return false;
    }
  };

  const handleSubmit = useCallback(async (data: any) => {
    const formIsValid = await validForm();

    if (formIsValid) {
      dispatch(setCreditCardInfo(data));
      // setToLocalStorage(BELEZA_NA_WEB_CREDIT_CARD, data);
      navigate('/cart/confirmation', { replace: true });
    }
  }, []);

  const handleChange = useCallback(
    (e: React.FormEvent<HTMLInputElement>) => {
      setPaymentInfo({
        ...paymentInfo,
        [e.currentTarget.id]: e.currentTarget.value
      });
    },
    [paymentInfo]
  );

  const handleInputFocus = useCallback(
    (e: React.FormEvent<HTMLInputElement>) => {
      setPaymentInfo({
        ...paymentInfo
        // focused: e.currentTarget.id
      });
    },
    [paymentInfo]
  );

  return (
    <Container>
      <Form ref={formRef} onSubmit={handleSubmit}>
        <h2>CARTÃO DE CRÉDITO</h2>
        <Content>
          <FormContent>
            <InputsContent>
              <fieldset>
                <label htmlFor="cardNumber">Número do cartão:</label>
                <Input
                  name="cardNumber"
                  type="text"
                  placeholder={CARD_NUMBER_PLACEHOLDER}
                  mask="creditCard"
                  defaultValue={creditCardInfo?.cardNumber || ''}
                  onChange={handleChange}
                  onFocus={handleInputFocus}
                  radius="all"
                />
              </fieldset>

              <fieldset>
                <label htmlFor="titularName">Nome do Titular:</label>
                <Input
                  name="titularName"
                  type="text"
                  placeholder={TITULAR_NAME_PLACEHOLDER}
                  mask="titularName"
                  defaultValue={creditCardInfo?.titularName || ''}
                  onChange={handleChange}
                  onFocus={handleInputFocus}
                  radius="all"
                />
              </fieldset>

              <FormGroup>
                <fieldset>
                  <label htmlFor="validate">Validade (mês/ano):</label>
                  <Input
                    name="validate"
                    type="text"
                    placeholder={DATE_PLACEHOLDER}
                    mask="date"
                    defaultValue={creditCardInfo?.validate || ''}
                    onChange={handleChange}
                    onFocus={handleInputFocus}
                    radius="all"
                  />
                </fieldset>

                <fieldset>
                  <label htmlFor="cvv">CVV:</label>
                  <Input
                    name="cvv"
                    type="text"
                    placeholder={CVV_PLACEHOLDER}
                    mask="cvv"
                    defaultValue={creditCardInfo?.cvv || ''}
                    onChange={handleChange}
                    onFocus={handleInputFocus}
                    radius="all"
                  />
                </fieldset>
              </FormGroup>
            </InputsContent>

            <CartContent>
              <Cards
                focused={paymentInfo?.focused || ''}
                cvc={paymentInfo?.cvv || ''}
                expiry={paymentInfo?.validate || ''}
                name={paymentInfo?.titularName || ''}
                number={paymentInfo?.cardNumber || ''}
              />
            </CartContent>
          </FormContent>

          {cartItems && <SumInfo cart={cartItems} />}
          <Button type="submit" title="Finalizar Pagamento" />
        </Content>
      </Form>
    </Container>
  );
}
