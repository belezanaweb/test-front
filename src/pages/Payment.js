import React from 'react'

const Payment = () => <h1>Payment</h1>

export default Payment
