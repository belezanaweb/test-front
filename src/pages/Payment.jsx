import React from 'react'

export default class Payment extends React.Component {
  state = {
    subTotal: '',
    shippingTotal: '',
    discount: '',
    total: '',
    ccn: '',
    name: '',
    validity: '',
    cvv: ''
  }

  componentDidMount() {
    this.fetchPrices()
  }

  handleChange = ({ target }) => {
    const { name } = target
    const value = target.type === 'checkbox' ? target.checked : target.value
    this.setState({
      [name]: value
    })
  }

  handleClick = () => {
    const { ccn, name, validity, cvv } = this.state
    const personInfos = {
      name,
      ccn,
      validity,
      cvv
    }
    localStorage.setItem('Infos', JSON.stringify(personInfos))
  }

  fetchPrices = () => {
    const myObj = JSON.parse(localStorage.getItem('Prices'))
    const { subTotal, shippingTotal, discount, total } = myObj
    this.setState({
      subTotal,
      shippingTotal,
      discount,
      total
    })
  }
  render() {
    const { subTotal, shippingTotal, discount, total, ccn, name, validity, cvv } = this.state
    return (
      <>
        Cartão de crédito
        <form>
          {/* Box das inputs */}
          <label htmlFor="ccn">
            Número do cartão:
            <input
              type="text"
              placeholder="XXXX-XXXX-XXXX-XXXX"
              id="ccn"
              maxLength="16"
              name="ccn"
              value={ccn}
              onChange={this.handleChange}
            />
          </label>
          <label htmlFor="name">
            Nome do titular:
            <input
              id="name"
              placeholder="Como no cartão"
              name="name"
              value={name}
              onChange={this.handleChange}
            />
          </label>
          <label htmlFor="validity">
            Validade:
            <input
              id="validity"
              placeholder="__/____"
              name="validity"
              value={validity}
              onChange={this.handleChange}
            />
          </label>
          <label htmlFor="cvv">
            CVV:
            <input id="cvv" placeholder="___" name="cvv" value={cvv} onChange={this.handleChange} />
          </label>
          {/* Box dos preços */}
          <div>
            <p>Produtos: R$ {subTotal}</p>
            <p>Frete: R$ {shippingTotal}</p>
            <p>Desconto: R$ {discount}</p>
            <p>Total: R$ {total}</p>
            <button type="button" onClick={this.handleClick}>
              Finalizar o pedido
            </button>
          </div>
        </form>
      </>
    )
  }
}
