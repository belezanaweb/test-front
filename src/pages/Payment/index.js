import React from 'react'

const Payment = () => {
  return (
    <div>
      <h1>Payment</h1>
    </div>
  )
}

export default Payment
