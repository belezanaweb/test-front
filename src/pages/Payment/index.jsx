import React, { useState } from 'react'
import { Container } from './styles'
import { useDispatch } from 'react-redux'
import { handlePayment } from '../../redux/slices/paymentSlice'

import loadable from '@loadable/component'

const Header = loadable(() => import('../../components/Header'), {
  resolveComponent: (components) => components.Header
})
const PaymentForm = loadable(() => import('../../components/PaymentForm'), {
  resolveComponent: (components) => components.PaymentForm
})
const Summary = loadable(() => import('../../components/Summary'), {
  resolveComponent: (components) => components.Summary
})
const ActionButton = loadable(() => import('../../components/ActionButton'), {
  resolveComponent: (components) => components.ActionButton
})

export const Payment = () => {
  const [cardNumber, setCardNumber] = useState('')
  const [name, setName] = useState('')
  const [expiryDate, setExpiryDate] = useState('')
  const [cvv, setCvv] = useState('')
  const dispatch = useDispatch()

  const handlePaymentInformation = () => {
    const payload = {
      cardNumber,
      name,
      expiryDate,
      cvv
    }
    dispatch(handlePayment(payload))
  }

  return (
    <>
      <Header />
      <Container>
        <PaymentForm
          fillCardNumber={setCardNumber}
          fillName={setName}
          fillCVV={setCvv}
          fillExpiryDate={setExpiryDate}
        />
        <Summary />
        <ActionButton
          handlePaymentInformation={handlePaymentInformation}
          pathToGo={'/confirmation'}
          actionText={'Finalizar o pedido'}
        />
      </Container>
    </>
  )
}
