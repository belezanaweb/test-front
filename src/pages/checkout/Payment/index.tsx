import styled from "styled-components";
import { Button } from "../../../components/Button";
import { Header } from "../../../components/Header";
import { useCartProducts } from "../../../services/hooks/cart/useCartProducts";
import { Title, Wrapper } from "../../../theme/styles";
import { CartInfo } from "../../../components/CartInfo";
import { PaymentForm } from "../../../components/PaymentForm";
import { ColumnType } from "../../../types";
import { MouseEvent } from "react";
import { useNavigate } from "react-router";
import { Loading } from "../../commom/Loading";
import { useForm } from "../../../services/hooks/form/useForm";

function Payment() {
  const navigate = useNavigate();
  const { data, isLoading } = useCartProducts();

  const { formRef, formData, handleFormDataChange, isFormTriggered, handleTriggerForm } = useForm();

  if (isLoading || !data) {
    return (
      <Loading />
    );
  }

  const { items, ...cartInfo } = data!;

  const handleSubmit = (e: MouseEvent<HTMLAnchorElement>) => {
    e.preventDefault();
    handleTriggerForm(true);

    for (const [key, value] of Object.entries(formRef.current)) {
      if (!value) {
        return;
      }
    }

    localStorage.setItem('paymentData', JSON.stringify(formData));
    navigate('/checkout/confirmation');
  }

  return (
    <Content>
      <Header />
      <Wrapper>
        <Title>Cartão de Crédito</Title>
        <Row>
          <Column flex={4}>
            <PaymentForm
              ref={formRef}
              isFormTriggered={isFormTriggered}
              handleFormDataChange={handleFormDataChange}
            />
          </Column>
          <Column flex={2}>
            <CartInfo {...cartInfo} />
            <Button
              to="/checkout/confirmation"
              onClick={handleSubmit}
            >
              Finalizar o pedido
            </Button>
          </Column>
        </Row>
      </Wrapper>
    </Content>
  );
}

export default Payment;

const Content = styled.main``;

const Row = styled.div`
  @media screen and (min-width: ${({ theme }) => theme.deviceSize.tabletMin}) {
    display: flex;
  }
`;

const Column = styled.div<ColumnType>`
  @media screen and (min-width: ${({ theme }) => theme.deviceSize.tabletMin}) {
    flex: ${({ flex }) => flex};

    &:first-child {
      margin-right: 10px;
    }

    &:last-child {
      margin-left: 10px;
    }
  }
`;
