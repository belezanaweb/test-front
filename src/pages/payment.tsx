import React from 'react';
import { useCallback } from 'react';
import { useHistory } from 'react-router-dom';

import { Button, Card, CreditCardFormController, CreditCardForm, Heading, Layout, Total } from '../components';
import { useCart, useCreditCard } from '../enhancers';

const Payment: React.FC = () => {
  const router = useHistory();
  const { data } = useCart();
  const { save } = useCreditCard();

  const handleSubmit = useCallback((payload: any) => {
    save(payload);
    router.push('/success');
  }, [router, save]);

  return (
    <Layout>
      <CreditCardFormController onSubmit={handleSubmit}>
        {() => (
          <>
            <Heading>Cartão de Crédito</Heading>
            <Card>
              <CreditCardForm />
            </Card>
            <Total data={data} />
            <Button fluid type="submit">Finalizar o Pedido</Button>
          </>
        )}
      </CreditCardFormController>
    </Layout>
  )
}

export default Payment;
