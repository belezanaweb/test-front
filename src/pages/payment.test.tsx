import React from 'react';
import { Provider } from "react-redux";
import { act, fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import axios from 'axios';

import mock from '../store/models/__mocks__/cart.json';

import store from '../store';

import App from '../App';
import Payment from './payment';

jest.mock('axios');

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
  useLocation: () => ({
    pathname: '/payment'
  }),
}));

jest.mock("../components/CreditCardForm/CreditCardForm", () => {
  return function CreditCardForm() {
    return (
      <div data-testid="entries">
        <span>Entries</span>
      </div>
    );
  };
});

afterEach(() => {
  jest.resetAllMocks();
});

describe('Payment Page', () => {
  it('should render payment page', async () => {
    axios.get = jest.fn().mockResolvedValue({ data: mock });

    const { container, getByTestId } = render(
      <Provider store={store}>
        <App>
          <Payment />
        </App>
      </Provider>
    );

    await Payment;

    await act(async () => {
      fireEvent.submit(getByTestId('form'))
    });

    expect(container).toMatchSnapshot();
  });
});
