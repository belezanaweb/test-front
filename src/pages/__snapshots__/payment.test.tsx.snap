// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment Page should render payment page 1`] = `
<div>
  <div
    class="sc-fznZeY dLKWGd"
  >
    <nav
      class="sc-fzplWN lctvBs"
    >
      <ol>
        <li
          class="sc-fznyAO jXJCuQ"
          title="Sacola"
        >
          <a>
            Sacola
          </a>
        </li>
        <li
          class="sc-fznyAO dKAYU"
          title="Pagamento"
        >
          <a>
            Pagamento
          </a>
        </li>
        <li
          class="sc-fznyAO jXJCuQ"
          title="Confirmação"
        >
          <a>
            Confirmação
          </a>
        </li>
      </ol>
    </nav>
    <div
      class="sc-fznKkj iQEdFx"
    >
      <form
        data-testid="form"
      >
        <span
          class="sc-AxgMl sc-fzpans lfXbBG"
        >
          Cartão de Crédito
        </span>
        <div
          class="sc-AxiKw sc-AxhUy hkJMki"
        >
          <div
            data-testid="entries"
          >
            <span>
              Entries
            </span>
          </div>
        </div>
        <div
          class="sc-fzokOt sc-fzqNqU ffIUde"
        >
          <div
            class="sc-fzoyTs iRdJNx"
          >
            <span
              class="sc-AxgMl sc-fzoNJl jvFjPf"
              id="subTotal"
            >
              Produtos
            </span>
            <span
              aria-labelledby="subTotal"
              class="sc-AxgMl sc-fzoNJl jvFjPf"
            >
              R$ 624,80
            </span>
          </div>
          <div
            class="sc-fzoyTs iRdJNx"
          >
            <span
              class="sc-AxgMl sc-fzoNJl jvFjPf"
              id="shippingTotal"
            >
              Frete
            </span>
            <span
              aria-labelledby="shippingTotal"
              class="sc-AxgMl sc-fzoNJl jvFjPf"
            >
              R$ 5,30
            </span>
          </div>
          <div
            class="sc-fzoyTs iRdJNx"
          >
            <span
              class="sc-AxgMl sc-fzoNJl dYPlFQ"
              id="discount"
            >
              Desconto
            </span>
            <span
              aria-labelledby="discount"
              class="sc-AxgMl sc-fzoNJl dYPlFQ"
            >
              -R$ 30,00
            </span>
          </div>
          <div
            class="sc-fzoXWK iKJWid"
          >
            <span
              class="sc-AxgMl sc-fzpmMD cHYbdw"
              id="total"
            >
              Total
            </span>
            <span
              aria-labelledby="total"
              class="sc-AxgMl sc-fzpmMD cHYbdw"
            >
              R$ 618,90
            </span>
          </div>
        </div>
        <button
          class="sc-AxhCb fdRirk"
          type="submit"
        >
          Finalizar o Pedido
        </button>
      </form>
    </div>
  </div>
</div>
`;
